### ISSUE TITLE
Type-checking with `inspect.isfunction()` results in a loss of all type constraints about a callable.

### ISSUE THREAD

**Issue by Finistere** [microsoft/pyright#3389]
**Describe the bug**
Given a callable with type constraints, doing a type check with `inspect.isfunction()` narrows the type to `FunctionType`. At this point, it loses all information about the constraints which it shouldn't as a function still is a `Callable`. However, I'm not sure what can be done. `FunctionType` doesn't accept any type arguments. So to me, it seems like either we lose information on the types or we lose the fact that it's a function and not "just" a callable.

**To Reproduce**
Code to typecheck:

```python
import inspect
from typing import Any, Callable, ParamSpec, TypeVar

F = TypeVar('F', bound=Callable[..., Any])
P = ParamSpec('P')
T = TypeVar('T')


def decorator(f: Callable[P, T]) -> Callable[P, T]:
    if not inspect.isfunction(f):
        raise TypeError()
    return f  # <- Error here


def decorator_v2(f: F) -> F:
    if not inspect.isfunction(f):
        raise TypeError()
    return f  # <- Error here

```
```
scratch_1.py
  scratch_1.py:12:12 - error: Expression of type "FunctionType" cannot be assigned to return type "(**P@decorator) -> T@decorator"
    Type "FunctionType" cannot be assigned to type "(**P@decorator) -> T@decorator" (reportGeneralTypeIssues)
  scratch_1.py:18:12 - error: Expression of type "FunctionType" cannot be assigned to return type "F@decorator_v2"
    Type "FunctionType" cannot be assigned to type "F@decorator_v2" (reportGeneralTypeIssues)
```

Interestingly, it works partially when using `callable()`:

```python
def decorator(f: Callable[P, T]) -> Callable[P, T]:
    if not (callable(f) or inspect.isfunction(f)):
        raise TypeError()
    return f  # <- Works !


def decorator_v2(f: F) -> F:
    if not (callable(f) or inspect.isfunction(f)):
        raise TypeError()
    return f  # <- still an Error here
```
```
scratch_1.py
  scratch_1.py:18:12 - error: Expression of type "((...) -> Any) | FunctionType" cannot be assigned to return type "F@decorator_v2"
    Type "((...) -> Any) | FunctionType" cannot be assigned to type "F@decorator_v2" (reportGeneralTypeIssues)
```

And everything works well when only using `callable()`:

```python
def decorator(f: Callable[P, T]) -> Callable[P, T]:
    if not callable(f):
        raise TypeError()
    return f  # <- Works !


def decorator_v2(f: F) -> F:
    if not callable(f):
        raise TypeError()
    return f  # <- Error here
```

**Expected behavior**
No errors, if possible?

**VS Code extension or command-line**
Command line, version `1.1.239` with Python `3.9.11`.

**Comment by Finistere** [comment]
That makes sense, thanks for the response! It's just a shame I can't express in Python that I want a function `def ...` and not any callable which includes types typically. Well, at least not if I want to manipulate type variables.
